### ISSUE TITLE
QUERY: interpolate.interp1d: `ValueError: Expect x to not have duplicates`

### ISSUE THREAD

**Issue by fisakhan** [scipy/scipy#21550]
# scipy-1.14.1
scipy.interpolate.interp1d(far_train, thresholds, kind='slinear') generates the following error:

> File "/home/ak/.virtualenvs/torch/lib/python3.10/site-packages/scipy/interpolate/_interpolate.py", line 397, in __init__
>     self._spline = make_interp_spline(xx, yy, k=order,
> 
>   File "/home/ak/.virtualenvs/torch/lib/python3.10/site-packages/scipy/interpolate/_bsplines.py", line 1385, in make_interp_spline
>     raise ValueError("Expect x to not have duplicates")
> 
> ValueError: Expect x to not have duplicates

The dimensions and values of far_train and thresholds are:

> (2000,)
> (2000,)
> far_train:  [0. 1. 1. ... 1. 1. 1.]
> thresholds: [-1.    -0.999 -0.998 ...  0.997  0.998  0.999]

Some people suggest scipy-1.8.1 can solve this problem but I don't want to downgrade to 1.8.1.

**Comment by lucascolley** [comment]
[The docs](https://docs.scipy.org/doc/scipy/reference/generated/scipy.interpolate.interp1d.html) say that (for 1-d `y`), the length of `y` must be equal to the length of `x`. Since `2000 != 10`, this seems like misuse of the function?

**Comment by fisakhan** [comment]
Sorry @lucascolley. They are both 2000. I edited the values in my question.

**Comment by lucascolley** [comment]
In the notes section of the docs:

> If the values in x are not unique, the resulting behavior is undefined and specific to the choice of kind, i.e., changing kind will change the behavior for duplicates.

**Comment by ev-br** [comment]
What Lucas said: duplicates in `x` are not supported.

Basically, it boils down to this: `interp1d` (and other interpolators) aim to construct a single-valued function, `f(x)` given the data, `x` and `y`. Whereas  in your example,

```
x = [0., 1., 1., ....]
y = [-1.,  -0.999,  -0.998, ...]
```

so what is the value of `f(x=1)`? Is it `-0.999` or `0.998` or something else entirely? _For your particular application_, you can probably somehow interpret these duplicates, but `interp1d` lacks the domain knowledge.
 In the face of this ambiguity, `interp1d(x, y, kind='slinear')` refuses to guess and errors out.

So to summarize, to use 1D interpolators from scipy, you'll need to come up with a way to preprocess your data so that interpolators only see ordered `x` arrays with no duplicates.

And no, downgrading to scipy 1.8 is unlikely to help. There are cases where downgrading is the right answer, but this is not one of them.
